ansible_base/rbac/models.py

```diff
@@ -0,0 +1,612 @@
+        object_id_field = cls._meta.get_field('object_id')
+        return cls.objects.filter(object_id__in=permission_qs.values_list(Cast('object_id', output_field=object_id_field)))
+
+    @classmethod
+    def visible_items(cls, user):
+        "This ORs querysets to show assignments to both UUID and integer pk models"
+        return cls._visible_items(RoleEvaluation, user) | cls._visible_items(RoleEvaluationUUID, user)
+
+    @property
+    def cache_id(self):
+        "The ObjectRole GenericForeignKey is text, but cache needs to match models"
+        return RoleEvaluation._meta.get_field('object_id').to_python(self.object_id)
+
+
+class AssignmentBase(CommonModel, ObjectRoleFields):
+    """
+    This uses some parts of CommonModel to save metadata like documenting
+    the user who assigned the permission and timestamp when it happened.
+    This caches ObjectRole fields for purposes of serializers,
+    both models are immutable, making caching easy.
+    """
+
+    object_role = models.ForeignKey('dab_rbac.ObjectRole', on_delete=models.CASCADE, editable=False, null=True)
+    object_id = models.TextField(
+        null=True, blank=True, help_text=_('Primary key of the object this assignment applies to, null value indicates system-wide assignment')
+    )
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE, null=True)
+    modified_on = None
+    created_on = models.DateTimeField(
+        default=timezone.now,  # Needed to work in migrations as a through field, which CommonModel can not do
+        editable=False,
+        help_text="The date/time this resource was created",
+    )
+    modified_by = None
+
+    class Meta:
+        app_label = 'dab_rbac'
+        abstract = True
+
+    def __init__(self, *args, **kwargs):
+        """
+        Because through models are created via a bulk_create, the save method is usually not called
+        to get around this, we populate the user model after initialization
+        """
+        super().__init__(*args, **kwargs)
+        if not self.id:
+            user = get_current_user()
+            if user:
+                # Hazard: user can be a SimpleLazyObject, so use id
+                self.created_by_id = user.id
+        # Cache fields from the associated object_role
+        if self.object_role_id and not self.object_id:
+            self.object_id = self.object_role.object_id
+            self.content_type_id = self.object_role.content_type_id
+            self.role_definition_id = self.object_role.role_definition_id
+
+    def save(self, *args, **kwargs):
+        if self.id:
+            raise RuntimeError(f'{self._meta.verbose_name.title()} model is immutable, use RoleDefinition.give_permission method')
+        # skip over CommonModel save because it would error due to missing modified_by and created_on
+        return super(CommonModel, self).save(*args, **kwargs)
+
+
+class RoleUserAssignment(AssignmentBase):
+    role_definition = models.ForeignKey(
```

Comment:
why do we need `role_definition` on both RoleUserAssignment and ObjectRole?

Reply:
Good question! The core unchangeable aspects are:
 - the `ObjectRole` model maps a `RoleDefinition` and an object via a GenericForeignKey
 - The assignment models are a through model between `ObjectRole` and the actor models of user & team

Everything else on the assignment models is cruft. You can call it "cached" data. It's not the source of truth. So you would say that `ObjectRole` is the source-of-truth about the `role_definition` field. Since an assignment has 1 and exactly 1 object role, there's no ambiguity and the assignment's role definition should be the object role's role definition. In code you see just that, it is literally outright copied.

Why? For the UI. For clients. At a point, the decision was made to not expose `ObjectRole` objects in the API. While we could keep it on the related `object_role`, this would create headache for filtering and serializers and whatnot.

Since the assignment models and the object-role models are all immutable, that makes this exceptionally easy to get away with.

Reply:
> Why? For the UI. For clients. 

perfect, make sense. thanks!